delira/training/abstract_trainer.py

```diff
@@ -1,31 +1,57 @@
 from abc import abstractmethod
 import logging
 import pickle
+import typing
 
 from batchgenerators.dataloading import MultiThreadedAugmenter
 
+from .predictor import Predictor
 from .callbacks import AbstractCallback
+from ..models import AbstractNetwork
 
+import numpy as np
+import os
+from tqdm import tqdm
 
 logger = logging.getLogger(__name__)
 
-KEYS_TO_GUARD = ["use_gpu",
-                 "input_device",
-                 "output_device",
-                 "_callbacks"]
 
-
-class AbstractNetworkTrainer(object):
+class AbstractNetworkTrainer(Predictor):
     """
-    Defines an abstract API for Network Trainers
+    Defines a Base API and basic functions for Network Trainers
 
     See Also
     --------
     :class:`PyTorchNetworkTrainer`
+    :class:`TfNetworkTrainer`
 
     """
 
-    def __init__(self, fold=0, callbacks=[]):
+    __KEYS_TO_GUARD = ["use_gpu",
+                       "input_device",
+                       "output_device",
+                       "_callbacks"]
+
+    def __init__(self,
+                 network: AbstractNetwork,
+                 save_path: str,
+                 losses: dict,
+                 optimizer_cls: type,
+                 optimizer_params: dict,
+                 train_metrics: dict,
```

Comment:
Maybe allow `train_dataset_metrics`?

Reply:
This was not allowed on purpose, because this would need the whole trainset to be cached, which is often not a good idea.

Reply:
For classification and detection it is only needed to cache the label and bounding boxes, which is pretty reasonable. Nevertheless, I get why it's not in there but I think it would be nice to have the option.

Reply:
@ORippler do you think, weh should have this option for the trainset as well?

Reply:
What benefit would `train_dataset_metrics` bring us?